src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp

```diff
@@ -171,20 +160,56 @@ namespace OpenMS
     log_model_threshold_ = threshold;
   }
 
+  std::vector<Size> EnzymaticDigestion::tokenise(const AASequence& protein) const
+  {
+    String s = protein.toUnmodifiedString();
+    boost::regex re(enzyme_.getRegEx());
+    boost::sregex_token_iterator i(s.begin(), s.end(), re, -1);
+    boost::sregex_token_iterator j;
+    std::vector<Size> positions;
+    Size pos = -1;
+    while (i != j)
+    {
+      AASequence pep = AASequence::fromString((String)(*i++));
+      pos += pep.size();
+      positions.push_back(pos);
+    }
+    return positions;
+  }
+
   bool EnzymaticDigestion::isCleavageSite_(
-    const AASequence& protein, const AASequence::ConstIterator& iterator) const
+    const AASequence& protein, const AASequence::ConstIterator& iterator, const std::vector<Size> positions) const
   {
-    switch (enzyme_)
+    if (enzyme_.getRegEx() == "()") // no cleavage
     {
-    case ENZYME_TRYPSIN:
-      if (use_log_model_)
+      return false;
+    }
+    else
+    {
+      if (!use_log_model_)
       {
-        if (*iterator != 'R' && *iterator != 'K') // wait for R or K
+        // naive digestion
+        if (iterator != protein.end() - 1)
+        {
+          return std::find(positions.begin(), positions.end(), iterator - protein.begin()) != positions.end();
+        }
+        else
+        {
+          return enzyme_.getRegEx().hasSubstring(iterator->getOneLetterCode()) && *iterator != 'P';
```

Comment:
Hmm I find the old implementation a bit confusing.
Usually I would expect to specify whether also semi cleavage products should be generated.
e.g.: for the string TFGQVVAR
```
TFGQVVAR
 FGQVVAR
  GQVVAR
   QVVAR
    VVAR
     VAR
      AR
TFGQVVA
TFGQVV
TFGQV
TFGQ
TFG
TF
```
which may be of interest in some cases like e.g. generating protein digests for search engines. This is currently not the case.

The other use case is to enforce semi / full matches to a protein database (as e.g. done in PeptideIndexer). Here the type of cleavage site might be of interest and is used in isValidProduct() to test if a peptide should be accepted based on if one wants to accept partial (semi) or full matches.

So these are two different, though related concepts.

@cbielow what is your take on this